streams/shared/src/main/scala/zio/stream/ZSink.scala

```diff
@@ -915,17 +915,51 @@ object ZSink extends ZSinkPlatformSpecificConstructors {
   /**
    * Drops incoming elements as long as the predicate `p` is satisfied.
    */
-  def dropWhile[In](p: In => Boolean)(implicit trace: Trace): ZSink[Any, Nothing, In, In, Any] =
-    new ZSink(ZPipeline.dropWhile(p).toChannel)
+  def dropWhile[In](p: In => Boolean)(implicit trace: Trace): ZSink[Any, Nothing, In, In, Any] = {
+    lazy val ch: ZChannel[Any, ZNothing, Chunk[In], Any, Nothing, Chunk[In], Unit] =
+      ZChannel.readWithCause(
+        in => {
+          val out = in.dropWhile(p)
+          if (out.nonEmpty)
+            ZChannel.write(out) *> ZChannel.unit
```

Comment:
I'm not sure I follow what's going on here. In the case that `out.nonEmpty == true`, won't this write the current Chunk and stop reading? Meaning that any subsequent chunks won't be read?

Reply:
exactly, when `out.nonEmpty` is true the sink actually succeeds (with a `unit`), however it must first write the leftovers to downstream.

once the sink succeeds it will no longer read from upstream, it is possible for other sinks to keep reading the leftovers + remaining upstream (achieved by composing sinks, ie. using `flatMap`)